OpenTelemetry (S3M) Databases apps (#1794)
SumoLogic/sumologic-documentation: docs/integrations/databases/opentelemetry/mysql-opentelemetry.md

-   Dashboards in the Metrics folder are based on MySQL metrics.
-   Dashboards in the Logs folder are based on MySQL logs from both the Error and Slow Query log formats.
    -   Dashboards based on the Error format:
        -   MySQL - Error Logs
        -   MySQL - Failed Logins
        -   MySQL - Replication
    -   Dashboard based on Slow Query format:
        -   MySQL - Slow Queries
-   MySQL General Health is based on logs from the MySQL Error log format.
-   MySQL Replication is based on logs from the MySQL Error log format.
-   MySQL Slow Queries is based on logs from the MySQL Slow Queries log format.

The MySQL app dashboards dependent on error logs are based on the message types ERROR, NOTE, Warning, and Info. For more details on the MySQL log file format, see <http://dev.mysql.com/doc/refman/5.5/en/server-logs.html>.

## Fields creation in Sumo Logic for Mysql

Following are the tags which will be created as part of MySql App install if not already present. 

- `db.cluster.name` - User configured. Enter a name to identify this MySql cluster. This cluster name will be shown in the Sumo Logic dashboards.
- `db.system` - Has fixed value of **mysql**

> - `sumo.datasource` - has fixed value of **mysql**

Reviewer: ```suggestion
- `sumo.datasource` - Has fixed value of **mysql**
```

- `db.node.name` - Has the value of host name of the machine which is being monitored

## Prerequisites

> For metric collection [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/mysqlreceiver#prerequisites) are the prerequisites.

Reviewer: ```suggestion
**For metric collection** [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/mysqlreceiver#prerequisites) are the prerequisites.
```

For log collection 

Configure MySQL to log to a local file(s). MySQL logs written to a log file can be collected via the Local File Source of a Sumo Logic Installed Collector. To configure the MySQL log file(s), locate your local my.cnf configuration file in the database directory.

1.  Open `my.cnf` in a text editor.
2.  Set the following parameters in the `[mysqld]` section:
```sql
[mysqld]
    log_error = /var/log/mysql/error.log
    slow_query_log=1
    slow_query_log_file = /var/log/mysql/mysql-slow.log
    long_query_time=2
    long_query_time=2
```

> [Error Logs](https://dev.mysql.com/doc/refman/5.7/en/slow-query-log.html). By default, error logs are enabled and are logged at file specified by the log_error key.

Reviewer: ```suggestion
[Error Logs](https://dev.mysql.com/doc/refman/5.7/en/error-log.html). By default, error logs are enabled and are logged at file specified by the log_error key.
```

> [Slow Query Logs](https://dev.mysql.com/doc/refman/5.7/en/slow-query-log.html). slow_query_log=1 enables logging of slow queries to the file specified by `slow_query_log_file`. Setting `long_query_time=2` will cause queries that take more than two seconds to execute to be logged. The default value of `long_query_time` is 10 seconds.

Reviewer: ```suggestion
[Slow Query Logs](https://dev.mysql.com/doc/refman/8.0/en/slow-query-log.html). slow_query_log=1 enables logging of slow queries to the file specified by `slow_query_log_file`. Setting `long_query_time=2` will cause queries that take more than two seconds to execute to be logged. The default value of `long_query_time` is 10 seconds.
```

> [General Query Logs](https://dev.mysql.com/doc/refman/5.7/en/query-log.html). We don't recommend enabling general_log for performance reasons. These logs are not used by the Sumo Logic MySQL App.

Reviewer: ```suggestion
[General Query Logs](https://dev.mysql.com/doc/refman/8.0/en/query-log.html). We don't recommend enabling general_log for performance reasons. These logs are not used by the Sumo Logic MySQL App.
```

1.  Save the `my.cnf` file.
2.  Restart the MySQL server: `sudo mysql.server restart`

## Collecting Logs, Metrics & installing App for MySQL

Here are the steps for collecting logs, metric and installing the app:

### Step 1 : Set up collector

:::note
If you want to use an existing OpenTelemetry Collector, you can skip this step by selecting the **Use an existing Collector** option.
:::

> If you want to create a new collector please select the **Add a new collector** option.

Reviewer: ```suggestion
To create a new Collector:
	
1. Select the **Add a new Collector** option.
2. Select the platform for which you want to install the Sumo OpenTelemetry Collector.
	
This will generate a command which can be executed in the machine which needs to get monitored. Once executed it will install the Sumo Logic OpenTelemetry Collector agent.
```

Select the platform for which you want to install the Sumo OpenTelemetry collector.

> This will generate a command which can be executed in the machine which needs to get monitored. Once executed it will install the Sumo Logic OpenTelemetry collector agent.  

Reviewer: ```suggestion
```

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/MySql-OpenTelemetry/MySql-Collector.png')} alt="Collector" />

### Step 2: Configure integration

In this step we will be configuring the yaml required for Mysql collection.

Below is the required input:

-   **Endpoint** - enter the url of the server which needs to be monitored
-   **User Name** - enter the Mysql username.
-   **Error File log Path** - enter the path to the error log file for your mysql instance.
-   **Slow Transaction file log path** - enter the path to the slow log file for your mysql instance.
-   **Tags** : `db.cluster.name`

> 

Reviewer: ```suggestion

You can add any custom fields which you want to tag along with the data ingested in sumo.

```

| parse regex field=query_block "# Bytes_sent:\s+(?<bytes_sent>\d*)\s+Tmp_tables:\s+(?<tmp_tables>\d*)\s+Tmp_disk_tables:\s+(?<temp_disk_tables>\d*)\s+Tmp_table_sizes:\s+(?<tmp_table_sizes>\d*)\n" nodrop | parse regex field=query_block "# Bytes_sent:\s+(?<bytes_sent>\d*)\n" nodrop | parse regex field=query_block "SET timestamp=(?<set_timestamp>\d*);(?:\\n|\n)(?<sql_cmd>[\s\S]*);" nodrop
| fields -query_block
| avg(query_time)  as avg_time,  sum(query_time)  as total_time,  min(query_time)  as min_time,  max(query_time)  as max_time,  avg(rows_examined)  as avg_rows_examined,  avg(rows_sent)  as avg_rows_sent,  avg(Lock_Time)  as avg_lock_time, count as frequency group  by sql_cmd, db.cluster.name
| sort by avg_time | limit  10
```

## Sample metric  query

This sample query is from the panel - FSync Op Count
```sql

> deployment.environment=* metric=mysql.operations operation=fsyncs db_cluster=* | sum

Reviewer: ```suggestion
sumo.datasource=mysql deployment.environment=* db.cluster.name=* db.node.name=* metric=mysql.operations operation=fsyncs  | sum
```

The Sumo Logic App for MySQL is a unified logs and metrics app that helps you monitor the availability, performance and resource utilization of MySQL database clusters. Preconfigured dashboards and searches provide insight into the health of your MySQL clusters, replication status, error logs, query performance, slow queries, Innodb operations, failed logins and error logs.

This App supports MySQL version 8.0

We use the OpenTelemetry collector for MySQL metric collection and for collecting MySQL logs.

The diagram below illustrates the components of the MySQL collection for each database server. OpenTelemetry collector runs on the same host as MySQL, and uses the [Mysql receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/mysqlreceiver) to obtain MySQL metrics, and the [Sumo Logic OpenTelemetry Exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/sumologicexporter) to send the metrics to Sumo Logic. MySQL logs are sent to Sumo Logic through a [filelog receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver).

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/MySql-OpenTelemetry/MySql-Schematics.png')} alt="Schematics" />

Reviewer 2: This image link is broken.

-   MySQL Replication is based on logs from the MySQL Error log format.
-   MySQL Slow Queries is based on logs from the MySQL Slow Queries log format.

The MySQL app dashboards dependent on error logs are based on the message types ERROR, NOTE, Warning, and Info. For more details on the MySQL log file format, see <http://dev.mysql.com/doc/refman/5.5/en/server-logs.html>.

## Fields creation in Sumo Logic for Mysql

Following are the [fields](https://help.sumologic.com/docs/manage/fields/) which will be created as part of MySql App install if not already present. 

- `db.cluster.name` - User configured. Enter a name to identify this MySql cluster. This cluster name will be shown in the Sumo Logic dashboards.
- `db.system` - Has fixed value of **mysql**
- `sumo.datasource` - Has fixed value of **mysql**
- `db.node.name` - Has the value of host name of the machine which is being monitored

## Prerequisites

**For metric collection** 
For metric collection [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/mysqlreceiver#prerequisites) are the prerequisites.

**For log collection** 

> Configure MySQL to log to a local file(s). MySQL logs written to a log file can be collected via the Local File Source of a Sumo Logic Installed Collector. To configure the MySQL log file(s), locate your local my.cnf configuration file in the database directory.

Reviewer 2: ```suggestion
Configure MySQL to log to a local file(s). MySQL logs written to a log file can be collected via the Local File Source of a Sumo Logic Installed Collector. To configure the MySQL log file(s), locate your local `my.cnf` configuration file in the database directory.
```

1.  Open `my.cnf` in a text editor.
2.  Set the following parameters in the `[mysqld]` section:
```sql
[mysqld]
    log_error = /var/log/mysql/error.log
    slow_query_log=1
    slow_query_log_file = /var/log/mysql/mysql-slow.log
    long_query_time=2
    long_query_time=2
```

[Error Logs](https://dev.mysql.com/doc/refman/8.0/en/error-log.html). By default, error logs are enabled and are logged at file specified by the log_error key.

[Slow Query Logs](https://dev.mysql.com/doc/refman/8.0/en/slow-query-log.html). slow_query_log=1 enables logging of slow queries to the file specified by `slow_query_log_file`. Setting `long_query_time=2` will cause queries that take more than two seconds to execute to be logged. The default value of `long_query_time` is 10 seconds.

[General Query Logs](https://dev.mysql.com/doc/refman/8.0/en/query-log.html). We don't recommend enabling general_log for performance reasons. These logs are not used by the Sumo Logic MySQL App.

1.  Save the `my.cnf` file.
2.  Restart the MySQL server: `sudo mysql.server restart`

> ## Collecting Logs, Metrics & installing App for MySQL

Reviewer 2: ```suggestion
## Collecting Logs, Metrics and installing App for MySQL
```

Here are the steps for collecting logs, metric and installing the app:

### Step 1 : Set up collector

:::note
If you want to use an existing OpenTelemetry Collector, you can skip this step by selecting the **Use an existing Collector** option.
:::

To create a new Collector:
	
1. Select the **Add a new Collector** option.
2. Select the platform for which you want to install the Sumo OpenTelemetry Collector.
	
This will generate a command which can be executed in the machine which needs to get monitored. Once executed it will install the Sumo Logic OpenTelemetry Collector agent.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/MySql-OpenTelemetry/MySql-Collector.png')} alt="Collector" />

Reviewer 2: This image link is broken.

### Step 2: Configure integration

In this step we will be configuring the yaml required for Mysql collection.

Below is the required input:

-   **Endpoint** - enter the url of the server which needs to be monitored
-   **User Name** - enter the Mysql username.
-   **Error File log Path** - enter the path to the error log file for your mysql instance.
-   **Slow Transaction file log path** - enter the path to the slow log file for your mysql instance.
-   **Tags** : `db.cluster.name`

You can add any custom fields which you want to tag along with the data ingested in sumo.

Click on the **Download YAML File** button to get the yaml file.

For linux platform - Click on **Download Environment Variables File** button to get the file with the password which is supposed to be set as environment variable.

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/MySql-OpenTelemetry/MySql-YAML.png')} alt="YAML" />

> ### Step 3: Sending logs and metric to Sumo

Reviewer 2: ```suggestion
### Step 3: Sending logs and metrics to Sumo
```

Once you have the yaml file downloaded in step 2, follow the below steps based on the platform of the machine :

<Tabs
  className="unique-tabs"
  defaultValue="Linux"
  values={[
    {label: 'Linux', value: 'Linux'},
    {label: 'Windows', value: 'Windows'},
    {label: 'macOS', value: 'macOS'},
  ]}>

<TabItem value="Linux">

1.  Copy the yaml file to `/etc/otelcol-sumo/conf.d/` folder in the MySql instance which needs to be monitored.

> 2.  Place Env file in the following directory

Reviewer 2: ```suggestion
2.  Place Env file in the following directory:
```

```sh
/etc/otelcol-sumo/env/
```

> 3.  restart the collector using

Reviewer 2: ```suggestion
3.  Restart the collector using:
```

```sh
 sudo systemctl restart otelcol-sumo
```

</TabItem>
<TabItem value="Windows">

1.  Copy the yaml file to `C:\ProgramData\Sumo Logic\OpenTelemetry Collector\config\conf.d` folder in the machine which needs to be monitored.

> 2.  Restart the collector using 

Reviewer 2: ```suggestion
2.  Restart the collector using:
```

```sql
sumo.datasource=mysql deployment.environment=* db.cluster.name=* db.node.name=* metric=mysql.operations operation=fsyncs  | sum
```

## Viewing MySQL Dashboards

### Overview

The **MySql - Overview** dashboard gives you an at-a-glance view of the state of your database clusters by monitoring key cluster information such as errors, failed logins, errors, queries executed, slow queries, lock waits, uptime and more.

Use this dashboard to:

-   Quickly identify the state of a given database cluster

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/MySql-OpenTelemetry/MySql-Overview.png')} alt="Overview" />

### Error Logs

The **MySql - Error Logs** dashboard provides insight into database error logs by specifically monitoring database shutdown/start events, errors over time, errors, warnings and crash recovery attempts.

Use this dashboard to:

-   Quickly identify errors and patterns in logs for troubleshooting
-   Monitor trends in error logs and identify outliers
-   Ensure that server start, server stop and crash recovery events are in line with expectations
-   Dashboard filters allow you to narrow a search for database cluster.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/MySql-OpenTelemetry/MySql-Error-Logs.png')} alt="Error Logs" />

Reviewer 2: This image link is broken.

### Failed Logins

The **MySql - Failed Logins** dashboard provides insights into all failed login attempts by location, users and hosts.

Use this dashboard to:

-   Monitor all failed login attempts and identify any unusual or suspicious activity

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/MySql-OpenTelemetry/MySql-Failed-Logins.png')} alt="Failed Logins" />

Reviewer 2: This image link is broken.

### Replication

The **MySql - Replication** dashboard provides insights into the state of database replication.

Use this dashboard to:

-   Quickly determine reasons for replication failures
-   Monitor replication status trends

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/MySql-OpenTelemetry/MySql-Replication.png')} alt="Replication" />

Reviewer 2: This image link is broken.

### Slow Queries

The **MySql - Slow Queries** dashboard provides insights into all slow queries executed on the database.

Note: Slow queries are queries that take 10 seconds or more to execute (default value is 10 seconds as per mysql configuration which can be altered) and excessive slow queries are those that take 15 seconds or more to execute.

Use this dashboard to:

-   Identify all slow queries
-   Quickly determine which queries have been identified as slow or excessive slow queries
-   Monitor users and hosts running slow queries
-   Determine which SQL commands are slower than others
-   Examine slow query trends to determine if there are periodic performance bottlenecks in your database clusters

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/MySql-OpenTelemetry/MySql-Slow-Queries.png')} alt="Slow Queries" />

Reviewer 2: This image link is broken.

### Performance and Resource Metrics

The **MySql - Performance and Resource Metrics** dashboard allows you to monitor the performance and resource usage of your database clusters.

Use this dashboard to:

-   Understand the behavior and performance of your database clusters
-   Monitor key operational metrics around threads running, innodb waits and locks.
-   Monitor query execution trends to ensure they match up with expectations
-   Dashboard filters allow you to narrow a search for a specific database cluster

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/MySql-OpenTelemetry/MySql-Performance-and-Resource-Metrics.png')} alt="Performance and Resource Metrics" />

Reviewer 2: This image link is broken.
